### PR TITLE
SAK-41502: Excusing an individual grade should be reflected in score's

### PR DESCRIPTION
**Pull request by victorGomollon**
[SAK-41502](https://jira.sakaiproject.org/browse/SAK-41502): Excusing an individual grade should be reflected in score's
Database changes: https://github.com/sakaiproject/sakai-reference/pull/90

### CODE REVIEW

File: edu-services/gradebook-service/hibernate/src/hibernate/org/sakaiproject/tool/gradebook/GradingEvent.hbm.xml
Comment (bjones86) [inline edu-services/gradebook-service/hibernate/src/hibernate/org/sakaiproject/tool/gradebook/GradingEvent.hbm.xml:23]: As with the reference PR, I believe this column should be not-nullable. Conceptually, there is only two states: grade is included, or grade is excluded. Eliminating the 3rd state (null) makes this feature cleaner, easier to understand, and better conforms to standard database design practises.

Comment (SedueRey) [reply edu-services/gradebook-service/hibernate/src/hibernate/org/sakaiproject/tool/gradebook/GradingEvent.hbm.xml:23]: Hello, @bjones86, we've been testing this PR and we would need a null value because you can modify a grade but not modifying the "excluding" state so it's needed to be null because on the controller excluded value is generated null because is not modified in the event.

Null is threated as it on line 1870 of GradebookServiceHibernateImpl. You can see it too GradeLogPanel.java, on line 79. This is what it was designed. If we need to remove null this should be entirely refactored.

What do you think?

Comment (bjones86) [reply edu-services/gradebook-service/hibernate/src/hibernate/org/sakaiproject/tool/gradebook/GradingEvent.hbm.xml:23]: Personally I feel that this should be refactored to remove null as a possible value. I don't think there is a use case for having these as nullable. What does null represent? Conceptually, the grade is either included (false) or excluded (true).

With regards to `GradebookServiceHibernateImpl:1870`:

```
public void saveGradeAndExcuseForStudent(final String gradebookUid, final Long gradableObjectId, final String studentUid,
			final String grade, final boolean excuse) {
```

In this case it is a primitive `boolean`, which cannot be null. Primitive booleans must be initialized as either `true` or `false`, so I don't think it's possible to have a null value here.

Long story short, I think the code should be refactored to eliminate null as a possible value here, as it doesn't provide any value. I'm sorry if this causes you more work  :(

If anyone disagrees with me, or has another opinion, feel free to chime in.

Comment (SedueRey) [reply edu-services/gradebook-service/hibernate/src/hibernate/org/sakaiproject/tool/gradebook/GradingEvent.hbm.xml:23]: @bjones86 , the way it works is not exactly this way, sorry. 

There are, right now, three scenarios in the UI (before this PR). You can exclude a grade, you can include a grade but you can also modify a grade **without changing the state** of included/excluded. In all this three scenarios the event is always the same. We are just mapping the UI behaviour to save consistent events. 

Right now, in the UI, you can do those three scenarios and you'll have the same event for all three. We are trying to avoid that (is weird for the instructor).

Maybe we can talk if the approach is the best way to do it but, for now, there are three cases. Null would be "not changed". 

Should we map it to an integer? Thanks.

**Comment by SedueRey** [comment]
@bjones86 @ern changes have been made. Thanks!

**Review by bjones86 — DISMISSED** [review]

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookService.java
Comment (ern) [inline edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookService.java:65]: This would make a great enum

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java
Comment (ern) [inline edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java:41]: With hibernate its best to use Objects that better map to the state of the database, so Integer makes more sense here.

**Comment by SedueRey** [comment]
@ern Jorge has updated this MR. Could you take a look?

**Comment by SedueRey** [comment]
@ern could you launch pipeline again? It failed because npm servers were down in that moment. Thanks!

**Review by ern — CHANGES_REQUESTED** [review]
Needs to be rebased

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookService.java
Comment (ern) [inline edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookService.java:61]: maybe 
GradingEventStatus {
GRADE_NONE(0)
GRADE_EXCLUDED(1)
GRADE_INCLUDED(2)
}

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java
Comment (ern) [inline edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java:55]: this.excluded = GradebookService.GradeEventState.GRADE_UPDATE_GRADE;

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java
Comment (ern) [inline edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java:41]: private GradingEventStatus excluded;

File: gradebookng/bundle/src/main/bundle/gradebookng.properties
Comment (fostersdesign) [inline gradebookng/bundle/src/main/bundle/gradebookng.properties:392]: The action the instructor takes is called "Excuse", so let's use this terminology:

`Score of <b>{1}</b> excused by {2}`

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbGradeLog.java
Comment (ern) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbGradeLog.java:46]: private final GradingEventStatus excluded;

File: gradebookng/bundle/src/main/bundle/gradebookng.properties
Comment (fostersdesign) [inline gradebookng/bundle/src/main/bundle/gradebookng.properties:393]: To match the above change, let's use:

`Score of <b>{1}</b> included by {2}`

**Review by fostersdesign — DISMISSED** [review]
I have a few suggested wording changes, please.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbGradeLog.java
Comment (ern) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbGradeLog.java:52]: this.status = ge.getStatus();

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java
Comment (ern) [inline edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java:106]: public GradingEventStatus getStatus();

File: edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java
Comment (ern) [inline edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingEvent.java:110]: public GradingEventStatus setStatus(GradingEventStatus status);

**Review by ern — DISMISSED** [review]

**Comment by SedueRey** [comment]
Oh, c'mon... Three months later? Really?

**Comment by fostersdesign** [comment]
> Oh, c'mon... Three months later? Really? 

Sorry, @SedueRey. I haven't been looking at PRs for a while and we discussed it at today's Core Team call, so that's when I noticed it.

**Comment by fostersdesign** [comment]
There are massive amounts of line changes in the commit now that look like whole files were rewritten. Please redo the commit so only the changes you've made are indicated. Thanks.

**Comment by SedueRey** [comment]
@fostersdesign we're working on that right now. Thanks!!! We have been played by Windows :D

**Comment by fostersdesign** [comment]
> @fostersdesign we're working on that right now. Thanks!!! We have been played by Windows :D

Okay. Thanks, @SedueRey!

**Review by juanjmerono — APPROVED** [review]

**Pull request merged**
